### ISSUE TITLE
Woocommerce Rounding Issue With Tax

### ISSUE THREAD

**Issue by loulemos** [woothemes/woocommerce#34529]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

1. All our products are entered without VAT (23% where I'm located), so a product that will cost a client 3.50€ is entered as 2.8455285 in the backoffice, this will make it display 3.50€ in the front-end. 
2. If I add 7 units to the cart the price will be 24.50€, Woocommerce has no issue here, the price displayed is 24.50€
4. However on another product priced at €24.23 - entered as 19.6950407 - when I add 7 units the price becomes 169.58€, when the price should be 169.61€ (3 cent diference.)

- This issue occurs even when I tick the "products are added with Tax"
- The "Round tax at subtotal level, instead of rounding per line" doesn't change anything.
- It's not solved by adding a price with only two digits in the backend, eg - instead of 2.8455285 (3.50€) I added 2.84 (3.49€), when I add 2 products to the cart instead of getting (3.49 x 2= 6.98€) I get 6.99€ - if I add 7, instead of getting (3.49 x 7=24.43) I get 24.45€ - a 2 cent difference for the client. 

- The only fix I have tried at the moment is to change the Tax Status to "None" and enter the price like I want eg - 2.83, so that when I add 7 units to the card I get a price of 19.81 which is correct. And when I enter a promo price, say 1.95€ and add 7 units to the cart I get 13.65€ which is correct. 

As an example here is a live store with this same issue: https://github.com/woocommerce/woocommerce/issues/24184#issuecomment-768104155 James has the exact issue I have.

### Expected behavior

The price that is displayed for the consumer (€24.23 in the case above) amount to the correct value when more than 2 units are added to the cart. 

When I add 7 units to the cart I get 169.58€

### Actual behavior

When the price should be 169.61€ (3 cent diference.)

### Steps to reproduce

1. Enter a product with the price 19.6950407 in the backoffice, this will make it display €24.23 in the front-end. 
2. Make sure the prices are exclusive of taxes and make the tax 23%
3. Add 1 unit to the cart, it should display €24.23
4. Now add another, it displays €48.45 instead of €48.46
5. You can try adding 5 more for a total of 7, the price displayed will be €169.58 instead of €169.61

### WordPress Environment

`
### WordPress Environment ###

WC Version: 6.8.2
REST API Version: ✔ 6.8.2
WC Blocks Version: ✔ 8.0.0
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.0.2
WP Multisite: –
WP Memory Limit: 700 MB
WP Debug Mode: –
WP Cron: –
Language: pt_PT
External object cache: –

### Server Environment ###

PHP Version: 7.4.30
PHP Post Max Size: 100 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.52.1
OpenSSL/1.0.2u

SUHOSIN Installed: –
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.8.2
WC Database Prefix: kuw_
Total Database Size: 1601.70MB
Database Data Size: 986.37MB
Database Index Size: 615.33MB
kuw_woocommerce_sessions: Data: 41.02MB + Index: 0.14MB + Engine InnoDB
kuw_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_woocommerce_order_items: Data: 6.52MB + Index: 2.52MB + Engine InnoDB
kuw_woocommerce_order_itemmeta: Data: 78.61MB + Index: 32.11MB + Engine InnoDB
kuw_woocommerce_tax_rates: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
kuw_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_woocommerce_payment_tokenmeta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
kuw_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ez_descontinuado: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
kuw_: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_actionscheduler_actions: Data: 238.31MB + Index: 234.83MB + Engine InnoDB
kuw_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_actionscheduler_logs: Data: 159.22MB + Index: 124.25MB + Engine InnoDB
kuw_adtribes_my_conversions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_aepc_custom_audiences: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_aepc_logs: Data: 0.19MB + Index: 0.00MB + Engine InnoDB
kuw_bv_activities_store: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_comments: Data: 9.52MB + Index: 9.05MB + Engine InnoDB
kuw_commercekit_searches: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_commercekit_waitlist: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
kuw_commercekit_wishlist: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_commercekit_wishlist_items: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_dgwt_wcas_index: Data: 4.02MB + Index: 0.13MB + Engine InnoDB
kuw_dgwt_wcas_invindex_cache: Data: 1.52MB + Index: 0.03MB + Engine InnoDB
kuw_dgwt_wcas_invindex_doclist: Data: 14.02MB + Index: 14.59MB + Engine InnoDB
kuw_dgwt_wcas_invindex_wordlist: Data: 1.38MB + Index: 1.42MB + Engine InnoDB
kuw_dgwt_wcas_tax_index: Data: 0.08MB + Index: 0.05MB + Engine InnoDB
kuw_dgwt_wcas_var_index: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_e_events: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_e_notes: Data: 0.02MB + Index: 0.17MB + Engine InnoDB
kuw_e_notes_users_relations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
kuw_e_submissions: Data: 0.02MB + Index: 0.27MB + Engine InnoDB
kuw_e_submissions_actions_log: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
kuw_e_submissions_values: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_import_detail_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_import_log_detail: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_import_postID: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mailpoet_custom_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_dynamic_segment_filters: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_feature_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_forms: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mailpoet_log: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mailpoet_mapping_to_external_entities: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_newsletters: Data: 1.50MB + Index: 0.03MB + Engine InnoDB
kuw_mailpoet_newsletter_links: Data: 1.52MB + Index: 1.81MB + Engine InnoDB
kuw_mailpoet_newsletter_option: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_newsletter_option_fields: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_newsletter_posts: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_newsletter_segment: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_newsletter_templates: Data: 7.14MB + Index: 0.00MB + Engine InnoDB
kuw_mailpoet_scheduled_tasks: Data: 0.39MB + Index: 0.31MB + Engine InnoDB
kuw_mailpoet_scheduled_task_subscribers: Data: 22.55MB + Index: 9.52MB + Engine InnoDB
kuw_mailpoet_segments: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_mailpoet_sending_queues: Data: 61.52MB + Index: 0.03MB + Engine InnoDB
kuw_mailpoet_settings: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_statistics_bounces: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mailpoet_statistics_clicks: Data: 0.16MB + Index: 0.22MB + Engine InnoDB
kuw_mailpoet_statistics_forms: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_statistics_newsletters: Data: 3.52MB + Index: 4.03MB + Engine InnoDB
kuw_mailpoet_statistics_opens: Data: 1.52MB + Index: 2.97MB + Engine InnoDB
kuw_mailpoet_statistics_unsubscribes: Data: 0.08MB + Index: 0.05MB + Engine InnoDB
kuw_mailpoet_statistics_woocommerce_purchases: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_mailpoet_stats_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_mailpoet_subscribers: Data: 1.52MB + Index: 1.78MB + Engine InnoDB
kuw_mailpoet_subscriber_custom_field: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_subscriber_ips: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mailpoet_subscriber_segment: Data: 1.52MB + Index: 0.66MB + Engine InnoDB
kuw_mailpoet_user_agents: Data: 0.30MB + Index: 0.06MB + Engine InnoDB
kuw_mailpoet_user_flags: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mclean_refs: Data: 0.27MB + Index: 0.00MB + Engine InnoDB
kuw_mclean_scan: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_mrw_cities: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mrw_orders: Data: 6.52MB + Index: 0.00MB + Engine InnoDB
kuw_mrw_ranges: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_mrw_taxes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_ms_snippets: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_newsletter: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_newsletter_emails: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_newsletter_sent: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_newsletter_stats: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_newsletter_user_logs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_nextend2_image_storage: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_nextend2_section_storage: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_nextend2_smartslider3_generators: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_nextend2_smartslider3_sliders: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
kuw_nextend2_smartslider3_sliders_xref: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_nextend2_smartslider3_slides: Data: 0.08MB + Index: 0.11MB + Engine InnoDB
kuw_omnisend_logs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_options: Data: 23.20MB + Index: 4.38MB + Engine InnoDB
kuw_postmeta: Data: 107.59MB + Index: 78.91MB + Engine InnoDB
kuw_posts: Data: 39.45MB + Index: 7.03MB + Engine InnoDB
kuw_prli_clicks: Data: 0.02MB + Index: 0.19MB + Engine InnoDB
kuw_prli_links: Data: 0.02MB + Index: 0.17MB + Engine InnoDB
kuw_prli_link_metas: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_rank_math_404_logs: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_rank_math_analytics_adsense: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_rank_math_analytics_ga: Data: 6.50MB + Index: 5.33MB + Engine InnoDB
kuw_rank_math_analytics_gsc: Data: 31.56MB + Index: 42.47MB + Engine InnoDB
kuw_rank_math_analytics_inspections: Data: 9.52MB + Index: 1.02MB + Engine InnoDB
kuw_rank_math_analytics_keyword_manager: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_rank_math_analytics_objects: Data: 1.52MB + Index: 0.38MB + Engine InnoDB
kuw_rank_math_internal_links: Data: 0.33MB + Index: 0.11MB + Engine InnoDB
kuw_rank_math_internal_meta: Data: 0.31MB + Index: 0.00MB + Engine InnoDB
kuw_rank_math_redirections: Data: 0.13MB + Index: 0.02MB + Engine InnoDB
kuw_rank_math_redirections_cache: Data: 0.08MB + Index: 0.02MB + Engine InnoDB
kuw_sgs_log_events: Data: 15.52MB + Index: 0.00MB + Engine InnoDB
kuw_sgs_log_visitors: Data: 0.22MB + Index: 0.00MB + Engine InnoDB
kuw_shortpixel_folders: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_shortpixel_meta: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_siq_sync: Data: 0.14MB + Index: 0.00MB + Engine InnoDB
kuw_smackcsv_file_events: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_smackuci_events: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_smush_dir_images: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_termmeta: Data: 0.23MB + Index: 0.25MB + Engine InnoDB
kuw_terms: Data: 0.11MB + Index: 0.13MB + Engine InnoDB
kuw_term_relationships: Data: 1.28MB + Index: 1.33MB + Engine InnoDB
kuw_term_taxonomy: Data: 0.11MB + Index: 0.13MB + Engine InnoDB
kuw_ultimate_csv_importer_acf_fields: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_ultimate_csv_importer_mappingtemplate: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_ultimate_csv_importer_media: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_ultimate_csv_importer_shortcode_manager: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_usermeta: Data: 20.47MB + Index: 14.52MB + Engine InnoDB
kuw_users: Data: 1.52MB + Index: 0.63MB + Engine InnoDB
kuw_wcpdf_invoice_number: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
kuw_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wc_customer_lookup: Data: 1.52MB + Index: 0.66MB + Engine InnoDB
kuw_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_wc_order_coupon_lookup: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
kuw_wc_order_product_lookup: Data: 6.52MB + Index: 7.56MB + Engine InnoDB
kuw_wc_order_stats: Data: 1.52MB + Index: 1.14MB + Engine InnoDB
kuw_wc_order_tax_lookup: Data: 1.52MB + Index: 0.45MB + Engine InnoDB
kuw_wc_product_attributes_lookup: Data: 0.06MB + Index: 0.05MB + Engine InnoDB
kuw_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wc_product_meta_lookup: Data: 1.44MB + Index: 0.84MB + Engine InnoDB
kuw_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wc_revolut_orders: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wdr_order_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wdr_order_item_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wdr_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfblockediplog: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfblocks7: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
kuw_wfconfig: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
kuw_wfcrawlers: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wffilechanges: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wffilemods: Data: 30.58MB + Index: 0.00MB + Engine InnoDB
kuw_wfhits: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
kuw_wfhoover: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wfissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_wfknownfilelist: Data: 10.52MB + Index: 0.00MB + Engine InnoDB
kuw_wflivetraffichuman: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wflocs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wflogins: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
kuw_wfls_2fa_secrets: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wfls_settings: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfnotifications: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfpendingissues: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
kuw_wfpklist_template_data: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfreversecache: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wfsnipcache: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
kuw_wfstatus: Data: 0.34MB + Index: 0.19MB + Engine InnoDB
kuw_wftrafficrates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wpfront_ure_options: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wpgb_cards: Data: 0.09MB + Index: 0.03MB + Engine InnoDB
kuw_wpgb_facets: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
kuw_wpgb_grids: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
kuw_wpgb_index: Data: 2.02MB + Index: 2.44MB + Engine InnoDB
kuw_wpns_backup_report: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wpns_files_scan: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
kuw_wt_iew_action_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wt_iew_cron: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wt_iew_ftp: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
kuw_wt_iew_mapping_template: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_bv_fw_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_bv_ip_store: Data: 0.22MB + Index: 0.25MB + Engine InnoDB
wp_bv_lp_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.34MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB

### Post Type Counts ###

ae_global_templates: 1
attachment: 4490
custom_css: 1
customize_changeset: 3
elementor_icons: 1
elementor_library: 27
mailpoet_page: 1
nav_menu_item: 204
oembed_cache: 5
page: 14
post: 1
product: 4164
product_variation: 6
revision: 4420
shop_coupon: 119
shop_order: 11588
wp_block: 23
wp_global_styles: 1
yith_price_rule: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (31) ###

FiboSearch - AJAX Search for WooCommerce (Pro): by FiboSearch Team – 1.19.0
BBQ Firewall: by Jeff Starr – 20220517
Code Snippets: by Code Snippets Pro – 3.2.0
CommerceGurus Commercekit: by CommerceGurus – 1.3.0
GTM4WP: by Thomas Geiger – 1.16.1
Elementor Pro: by Elementor.com – 3.7.3
Elementor: by Elementor.com – 3.7.2
Eupago for WooCommerce: by WebAtual – 2.1.0.1
Flexible Shipping PRO: by Octolize – 2.12.1
Flexible Shipping: by Octolize – 4.13.3
Heartbeat Control by WP Rocket: by WP Rocket – 2.0
Kadence Blocks – Gutenberg Blocks for Page Builder Features: by Kadence WP – 2.4.13
Kadence WooCommerce Email Designer: by Kadence WP – 1.5.4
Kirki Customizer Framework: by David Vongries – 4.0.24
Loco Translate: by Tim Whitlock – 2.6.2
NIF (Num. de Contribuinte Português) for WooCommerce: by PT Woo Plugins (by Webdados) – 4.3.0
Perfect Brands for WooCommerce: by QuadLayers – 2.1.6
Portugal States (Distritos) for WooCommerce: by PT Woo Plugins (by Webdados) – 3.1.0
Rank Math SEO PRO: by Rank Math – 3.0.19
Rank Math SEO: by Rank Math – 1.0.96
Two Factor: by Plugin Contributors – 0.7.1
User Role Editor: by Vladimir Garagulya – 4.63
Product Feed ELITE for WooCommerce: by AdTribes.io – 2.3.5
Payment Plugins for Stripe WooCommerce: by Payment Plugins
support@paymentplugins.com – 3.3.26
Pay for Payment for WooCommerce: by Karolína Vyskočilová – 2.1.7
WooCommerce PayPal Payments: by WooCommerce – 1.9.2
WooCommerce: by Automattic – 6.8.2
WP Grid Builder - Elementor: by Loïc Blascos – 1.1.2
WP Grid Builder: by Loïc Blascos – 1.6.8
YITH WooCommerce Role Based Prices Premium: by YITH – 1.15.0

### Dropin Plugins (1) ###

object-cache.php: Object Cache Pro (Drop-in)

### Must Use Plugins (1) ###

Object Cache Pro (MU): by Rhubarb Group – 1.14.5

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)
woosb (woosb)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –

### WC Pages ###

Shop base: #6525 - /shop/
Cart: #6526 - /cart/
Checkout: #6527 - /checkout/
My account: #6528 - /minha-conta/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Shoptimizer Child Theme
Version: 1.2.1
Author URL: 
Child Theme: ✔
Parent Theme Name: Shoptimizer
Parent Theme Version: 2.6.0
Parent Theme Author URL: https://www.commercegurus.com/
WooCommerce Support: ✔

### Templates ###

Overrides: –

### WooCommerce PayPal Payments ###

Onboarded: ✔
Shop country code: PT
WooCommerce currency supported: ✔
PayPal card processing available in country: –
Pay Later messaging available in country: –
Webhook status: –
Vault enabled: ✔
Logging enabled: –
Reference Transactions: –
Used PayPal Checkout plugin: ✔

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
experimental-product-tour
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
wc-pay-promotion
wc-pay-welcome-page
wc-pay-subscriptions-page

Disabled Features: minified-js
settings

Daily Cron: ✔ Next scheduled: 2022-09-01 09:16:47 +01:00
Options: ✔
Notes: 70
Onboarding: completed

### Action Scheduler ###

Canceled: 7
Oldest: 2022-08-01 10:15:19 +0100
Newest: 2022-08-01 10:15:19 +0100

Complete: 718,882
Oldest: 2022-07-31 10:36:46 +0100
Newest: 2022-08-31 12:56:12 +0100

Failed: 4
Oldest: 2022-06-01 19:25:43 +0100
Newest: 2022-06-01 19:25:43 +0100

Pending: 3
Oldest: 2022-09-01 06:45:11 +0100
Newest: 2022-09-04 10:15:49 +0100

### Status report information ###

Generated at: 2022-08-31 13:06:21 +01:00
`

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by loulemos** [comment]
Here is a video to show what the clients see: 


https://user-images.githubusercontent.com/51758046/187708320-677bc80b-697a-4fab-a8dc-b8a53382b809.mp4



**As you can see:**
I first add 1 product and the price in the cart is €24.23 - **Correct**
Then I add another one, so now we have 2 products and the price in the cart is - €48.45 - **Incorrect** - it should be €48.46
Then I keep adding products until I have 7 products total in the cart, that sould amount to €169,61 - instead we get €169,58 - **a 3 cent difference.**

**And here is the exact situation I'm facing posted by another user with no solution**
Woocommerce thread that was closed: https://wordpress.org/support/topic/woocommerce-rounding-issue-with-tax/ - @shaunkuschel @conschneider
Stackoverflow thread with no answers: https://stackoverflow.com/questions/69871743/woocommerce-rounding-issue-with-tax

Some poor fella paid for a fix it seems:
https://www.freelancer.com/projects/woocommerce/woocommerce-tax-rounding-issue

Lads sorry to be annoying but we want to launch some big features on our site and we can't proceed until this is fixed. 

I found this thread https://github.com/woocommerce/woocommerce/pull/14475 from 2016 and there are comments from a month ago asking if there's a fix.

**Comment by samueljseay** [comment]
Thank you for all the detail on this @loulemos We will look at what we can do to address this very soon.

**Comment by loulemos** [comment]
Good morning everyone, thanks a lot @samueljseay 
I'll be following this thread closely, have a good day.

**Comment by vedanshujain** [comment]
Hey @loulemos, as a general rule you would need to display prices as they are entered to prevent rounding issues. For example, in your case, since you are displaying prices including taxes, you should enter them including taxes as well. This mismatch in setting is causing the apparent rounding issue.


In my test site, I added a product with a price 24.23 with 23% tax, and this is how it's displayed:
<img width="1193" alt="Screenshot 2022-09-07 at 9 33 10 PM" src="https://user-images.githubusercontent.com/7571618/188925917-e294e622-52ca-4eaa-be13-cd7a76c3916a.png">

On the same site, this is how products with price 3.5 is displayed:

<img width="1180" alt="Screenshot 2022-09-07 at 9 34 13 PM" src="https://user-images.githubusercontent.com/7571618/188926163-4624a1ae-9780-44cc-9f35-c609da76b959.png">

This is the edit product page:

<img width="906" alt="Screenshot 2022-09-07 at 9 38 47 PM" src="https://user-images.githubusercontent.com/7571618/188927131-02355115-476f-403f-8b9e-9b78fef7b8a6.png">


For reference these are my settings:

Taxes:

<img width="1317" alt="Screenshot 2022-09-07 at 9 34 47 PM" src="https://user-images.githubusercontent.com/7571618/188926256-171dee13-7c9f-4b8e-8188-f03b79a0a3ac.png">

Tax display settings:

<img width="961" alt="Screenshot 2022-09-07 at 9 35 17 PM" src="https://user-images.githubusercontent.com/7571618/188926361-385d8125-8cfe-4519-87be-72dfde99a4c7.png">

In your case, you are seeing the rounding issues because you are entering prices before taxes but are displaying them including it. The root cause is that WooCommerce will use whatever you have entered while doing the calculations, which means if you have entered prices, including taxes, it will figure out the gross (incl. tax) price once and then calculate tax from the gross. Similarly, if you enter the price excluding taxes, WC will figure out the net (excl. tax) price and then calculate taxes on top of this net.

This is not a bug, but is intended to support both modes of entering prices (net and gross). Whenever tax amount cannot be perfectly represented in two decimals, WC will have do some rounding when prices are displayed, and therefore an inaccuracy is induced. This inaccuracy is presented to the user when the mode of displaying prices is different the mode if entering it (for example, entering net but displaying gross).

That said, it's entirely possible that I am not able to understand your issue and something indeed needs to be fixed. However, if I understand correctly, if you change the product price and setting to enter them, including taxes, this rounding error will go away, and you won't have to do any complicated calculations to figure out prices with high precision.

**Comment by joashrajin** [comment]
@vedanshujain we have another report of this in #5571664-zen. I am noticing something similar:

They have ten products worth 109,44 € (including tax). On the checkout page, the price is showing as 109,44 € * 8 = 875,54 € (ex. VAT).

When I manually calculate  109,44 € * 8, it should be 875.52 €. Any idea why we see a small discrepancy in the decimals?


Tax Settings: 
<img width="708" alt="image" src="https://user-images.githubusercontent.com/16253818/191459499-b0696d37-3558-4f90-9057-a22e61b1b983.png">
<img width="1601" alt="image" src="https://user-images.githubusercontent.com/16253818/191459622-6af38847-caa3-46f5-9917-e2379bf224d1.png">

**Comment by studiograficorosati** [comment]
Hi I have the same problem too,

My duplicate was closed rightly, but I want to share my experience ...

I have a big problem with decimals, I applied the Italian VAT at 22% but when you calculate the product in the cart the decimals are wrong.

Example:
10 products in the cart from € 133.68 (22% VAT excluded) exit at € 1,336.77 when in reality 133.68 × 10 = 1,336.80

I have read many forums and many people have this problem.

It has been an unsolved problem for several years. Is it possible that Woocommerce can't solve it?!?!?

The solution is not to remove the VAT as the site is dedicated to retailers and it is unthinkable that they see prices including VAT.

There are no online solutions that fix the problem.

I represent a company that has a turnover of millions of euros. It is not a problem of budget but of credibility. Prices, even if only a few cents, are always wrong.

I got one of the best paid programmers to try to fix the bug, here is his answer:

“… .. I have dedicated many days and hours to this project and in that time I can even build a complete system…. however Woocommerce does not provide any fixes for this bug.

I created all the code from scratch ...

Unfortunately, its bug is in woocommerce core files which saves data in 16-digit format ...

It is not possible to get correct decimals in this version of woocommerce. you can convince with the customer unfortunately his fault is not in our hands.

It is not possible to check 16 digits after the period.

I've tried everything and nothing happened nobody can do it, I can bet.

I've tried everything ...
![Schermata 2022-09-21 alle 10 23 11](https://user-images.githubusercontent.com/114060858/191462190-8f2b1c32-3843-413f-98d5-9b6f4a4f0ae4.png)

it is useless if we make changes to the plugin if it then disappears every time you update Woocommerce, because the core files all change.

The only smart way is to wait for the next Woocommerce updates that will address the problem ...

Hundreds of questions on the woocommce page regarding this bug ... `"

the solution is not even postrtare to 4/5 decimal place
![Schermata 2022-09-21 alle 10 23 52](https://user-images.githubusercontent.com/114060858/191462184-5027edc3-cd96-47ab-8104-f33d754c7ef1.png)
s because in Italy there are 2 ...

**Comment by vedanshujain** [comment]
Hey @joashrajin, 

> When I manually calculate 109,44 € * 10, it should be 875.52 €. 

Shouldn't this be 897.05 (calculated from 1094.4 / 1.22), as the total price is 1094.4 and the tax is 22%? Anycase, since we recommend displaying prices as they are entered, you should set the setting `Display prices in the shop` and `Display prices in the basket and checkout to `Including Tax`. This is what I see on my testing site with this configuration:

<img width="1146" alt="Screenshot 2022-09-21 at 3 35 36 PM" src="https://user-images.githubusercontent.com/7571618/191477245-c96ab185-907a-4fc2-88cc-24c858e86737.png">

Hey @studiograficorosati 

If I understand correctly, you want to enter prices excluding taxes, and then have it calculated later on. This is entirely possible without any rounding issues, on my test site:

For product with price 133.68:

<img width="1076" alt="Screenshot 2022-09-21 at 3 43 37 PM" src="https://user-images.githubusercontent.com/7571618/191479272-8f47998e-21fd-423e-8600-4a4286644b48.png">

And then for product with price 4.38

<img width="1078" alt="Screenshot 2022-09-21 at 3 45 01 PM" src="https://user-images.githubusercontent.com/7571618/191479352-644cd05d-2656-43b4-8ff6-fc628919bf01.png">

 I am entering both prices excluding taxes, see:

<img width="779" alt="Screenshot 2022-09-21 at 3 47 07 PM" src="https://user-images.githubusercontent.com/7571618/191479472-613018ba-c530-420a-8b96-08785f6360f9.png">

and 

<img width="910" alt="Screenshot 2022-09-21 at 3 47 19 PM" src="https://user-images.githubusercontent.com/7571618/191479509-8d7bb2ac-5da8-448d-8f8e-c805301ccc2d.png">


These are my tax settings:

<img width="718" alt="Screenshot 2022-09-21 at 3 47 49 PM" src="https://user-images.githubusercontent.com/7571618/191479619-ffb154ad-c044-4feb-9f13-b58328e6c406.png">

And tax is configured at 22%:

<img width="1311" alt="Screenshot 2022-09-21 at 3 48 13 PM" src="https://user-images.githubusercontent.com/7571618/191479698-f9c8394e-d2a9-4f6b-8da7-b4d880ea88a9.png">

Can you confirm if you are trying with these settings? It could be that I am not understanding the issue correctly, in which case screenshot of settings page like above might be helpful.

**Comment by studiograficorosati** [comment]
Hello,

unfortunately the management system includes prices including VAT, ....

Are you telling me that to solve the problem you hav
![Schermata 2022-09-21 alle 12 31 49](https://user-images.githubusercontent.com/114060858/191482712-fa505f6e-cac6-4527-975e-92b22b42d946.png)
![Schermata 2022-09-21 alle 12 31 32](https://user-images.githubusercontent.com/114060858/191482715-e118553e-edfc-49bb-a34a-4c4ceca6e1c3.png)
![Schermata 2022-09-21 alle 12 31 20](https://user-images.githubusercontent.com/114060858/191482718-96f4f326-7935-4ab7-8496-01554311786f.png)
![Schermata 2022-09-21 alle 10 23 52](https://user-images.githubusercontent.com/114060858/191482719-bbb71207-efd8-42f2-b555-c806d4dbeb43.png)
e to enter the prices excluding VAT?
![Schermata 2022-09-21 alle 12 32 20](https://user-images.githubusercontent.com/114060858/191482709-9a9bbfb2-ba56-4ad9-9d69-239f650beed1.png)

**Comment by joashrajin** [comment]
@vedanshujain thanks for the update. I made a small typo, I meant to say 109.44 x 8 (not x 10). I have updated the message. Just to clarify, the issue I reported is the same one as @studiograficorosati

**Comment by vedanshujain** [comment]
Hey,

> Are you telling me that to solve the problem you have to enter the prices excluding VAT?

No, not necessary, although it would solve the problem as well. You need to display the prices as they are entered, which means if you are entering them inclusive of taxes, you should display them inclusive of taxes. In this case, this is how they would be displayed:

<img width="1062" alt="Screenshot 2022-09-21 at 4 35 52 PM" src="https://user-images.githubusercontent.com/7571618/191488696-4ec218b4-dd30-403b-8541-94d3147b8bf0.png">

Note that the "subtotal" is configurable using the `woocommerce_order_subtotal_to_display` so you can change it to say `$534.00 (includes 96.30 tax)` or something similar if you want.

The cause is that WC will try to honor prices as they are entered and optimize calculation towards them by calculating the total first and then computing taxes. Computation of taxes is where rounding generally happens, so if you enter prices inclusive of taxes but display prices excluding them, your customer may see the rounding issue since they are more likely to see a rounded value. Whereas if you display prices as they are entered, your customers will see a price that is effectively price_entered * quantity, thus eliminating any rounding component from being displayed.

**Comment by vedanshujain** [comment]
This is my setting for above: 

<img width="884" alt="Screenshot 2022-09-21 at 4 44 57 PM" src="https://user-images.githubusercontent.com/7571618/191490159-54c8ab6b-7298-45bc-a6d8-790051ff0088.png">

**Comment by studiograficorosati** [comment]
Unfortunately, it is a site dedicated to resellers that cannot display the prices including taxes, the only way to solve the problem is to ask the management system if the prices are loaded in Woocommerce excluding taxes ...

**Comment by nodemonsdev** [comment]
Hi! I had this problem also, we solved it by disabling the woocommerce tax calculation and calculating it from the product price when and where needed manually (custom calculation functions to display the tax)... this also involved changing the API response when the clients bookkeeping platform asked for orders.

I think most of the rounding problem comes from woocommerce using the subtotal of a cart item line or order line to calculate tax instead of calculating tax for each item's price and then summing the tax for the total tax. Its more logical that EACH product's price contains tax than the order line subtotal.

**Comment by studiograficorosati** [comment]
Thanks @nodemonsdev for your testimony, but we cannot do things manually, we have a management software that synchronizes the products, the only thing we can do is ask the management to send the prices excluding VAT in order to avoid the problem of decimals. ..

**Comment by nodemonsdev** [comment]
@studiograficorosati I did not mean manually as in calculating the price yourself. We did custom calculation functions where we needed to display the tax.

**Comment by asdfMaciej** [comment]
Hello,

I've encountered this issue as well when storing brutto prices. This should be configurable in WooCommerce.
The issue shines most when dealing with large quantities of low-value items (try 2000 items with 4,224 rounded to 4,22). 

I've created this snippet which partially fixes the issue:
```php
add_filter( 'woocommerce_get_price_excluding_tax', 'round_price_to_2_decimal_places',10,3);

function round_price_to_2_decimal_places($price, $qty, $product) {
    return round($price / $qty, 2) * $qty;
}
```

There is still a slight mismatch, but smaller by a few orders of magnitude. Case in point:
- item A, 4,45 PLN brutto per unit, quantity 1000
- item B, 4,45 PLN, quantity 1000
- item C, 4,56 PLN, quantity 1000

Total: 13460,00 PLN
The cart preview displays it as 13460,00 PLN, including 2516,90 PLN of 23% tax.

13460,00 PLN brutto actually includes 2516,91 PLN of tax. This is a slight error (0,01 PLN), but still an accounting error. Hope that someone finds an improved patch until WooCommerce devs tackle it. Perhaps the calculation should be modified in other places.

**Comment by vedanshujain** [comment]
Hey @asdfMaciej Can you share bit more details about the tax setting configuration you have like it was shared [in this comment](https://github.com/woocommerce/woocommerce/issues/34529#issuecomment-1253515963)? I would love to help you to set up (or fix) WC in a way that there are no rounding issues.

**Comment by namiokuzono** [comment]
7321144-zen

**Comment by pierorocca** [comment]
Hi @loulemos - thanks for reporting this, and for the detail you gave.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I checked that tax calculation and it's working as expected. Where the rounding happens is going to vary based on the different settings and tax approaches. Early rounding can cause a "lie" in the shopper's price. The higher the quantity rounded, the more pronounced the impact.

For the EU, the foolproof approach is to enter prices inclusive of tax. 

Here's how where rounding occurs can create different answers.

For example: 2.84 x 2 = 5.68. Then calculate tax 5.68 X 1.23 = 6.9864 which then rounds up to 6.99.

Your calculation: 2.84 * 1.23 = 3.4932.  That rounds down to 3.49. Then 3.49 x 2 = 6.98.

At a qty of 10, the gap increases to 3 cents.

Closing this issue.
